### PR TITLE
Hank's live stacking transactions

### PR DESCRIPTION
**Pull request by hstove**
> [Download the latest builds](https://github.com/blockstack/stacks-wallet/actions/runs/315690823)

This builds on top of #262 . I'm making a separate PR so that Kyran can more easily code review my work.

### CODE REVIEW

File: app/modals/stacking/stacking-modal.tsx [hunk 1]
```diff
@@ -37,13 +48,22 @@ enum StackingModalStep {
   NetworkError,
 }
 
+const poxClient = new POX(DEFAULT_STACKS_NODE_URL);
```
Comment (kyranjamie) [inline app/modals/stacking/stacking-modal.tsx:51]: This needs to be pulled in from the `selectedNodeApi` selector, so it updates when user changes node.

File: app/modals/stacking/stacking-modal.tsx [hunk 2]
```diff
@@ -56,47 +76,87 @@ export const StackingModal: FC<StackingModalProps> = ({ onClose }) => {
...
       salt,
       password,
     });
-    console.log({ privateKey });
-    // return makeSTXTokenTransfer({ ...options, senderKey: privateKey });
-  }, [encryptedMnemonic, password, salt]);
+    const txOptions = await poxClient.getLockTxOptions({
+      cycles: numCycles,
+      poxAddress,
+      amountSTX: poxInfo.min_amount_ustx,
```
Comment (kyranjamie) [inline app/modals/stacking/stacking-modal.tsx:107]: Scope for confusion; isn't clear this prop refers ustx

File: app/utils/stacking/pox.ts [hunk 3]
```diff
@@ -25,20 +28,38 @@ interface POXInfo {
...
     version: Buffer;
     hashbytes: Buffer;
   };
   btcAddress: string;
 }
 
+interface StackerInfoCV {
+  'amount-ustx': UIntCV;
+  'first-reward-cycle': UIntCV;
+  'pox-addr': {
```
Comment (kyranjamie) [inline app/utils/stacking/pox.ts:45]: Hypen-based naming used elsewhere spreads like a highly infectious, deadly illness. Is this something we just have to live with given clarity's const naming 🤔

File: app/utils/sum-stx-tx-total.ts [hunk 4]
```diff
@@ -11,6 +12,15 @@ export function sumStxTxTotal(address: string, tx: Transaction) {
     return new BigNumber(tx.fee_rate);
   }
 
+  if (isLockTx(tx) && tx.tx_result?.repr) {
+    // We get the amount stacked from the tx_result
+    const matcher = /\(tuple \(lock-amount\su(\d+)/;
+    const matches = matcher.exec(tx.tx_result.repr);
+    if (matches && matches[1]) {
+      return parseInt(matches[1]);
```
Comment (kyranjamie) [inline app/utils/sum-stx-tx-total.ts:20]: Better to use BigNumber, avoid possible missing radix value bugs?

File: app/utils/tx-utils.ts [hunk 5]
```diff
@@ -9,3 +9,7 @@ export function getRecipientAddress(tx: Transaction) {
   if (tx.tx_type !== 'token_transfer') return null;
   return tx.token_transfer.recipient_address;
 }
+
+export function isLockTx(tx: Transaction): boolean {
+  return tx.tx_type === 'contract_call' && tx.contract_call.contract_id.includes('pox');
```
Comment (kyranjamie) [inline app/utils/tx-utils.ts:14]: Can other people submit contracts with pox in?

**Review by kyranjamie — COMMENTED** [review]
Just back from Italy, had a quick pass through, looks good 👍 

We do need to give a thought to how to best handle default/not loaded yet balances. Ideally should avoid inline `balance || 0` checks (which I also did).

File: app/utils/stacking/pox.ts (hunk 3 above)
Comment (hstove) [reply app/utils/stacking/pox.ts:45]: lol, agreed. In this case we'll need the hyphens no matter what for this type signature. When auto-generated clarity clients with camelcase conversion??

File: app/modals/stacking/stacking-modal.tsx (hunk 2 above)
Comment (hstove) [reply app/modals/stacking/stacking-modal.tsx]: :+1: changed to `amountMicroSTX`

File: app/utils/tx-utils.ts (hunk 5 above)
Comment (hstove) [reply app/utils/tx-utils.ts]: Good catch, yes they can. I've refactored to match the exact contract ID of the pox contract.

**Comment by hstove** [comment]
@kyranjamie I've updated the stacking logic so that it uses your entire balance (minus a fee). I also tweaked some things on my local mocknet, which fixed the failed stacking TXs I was seeing. It seems like the whole 'pending' flow works properly when the TX is successful.

File: app/modals/stacking/stacking-modal.tsx [hunk 6]
```diff
@@ -56,47 +71,96 @@ export const StackingModal: FC<StackingModalProps> = ({ onClose }) => {
...
+      const resp: ResponseSign = await blockstackApp.sign(
+        `m/44'/5757'/0'/0/0`,
+        unsignedTx.serialize()
+      );
+      if (resp.returnCode !== LedgerError.NoErrors) {
+        throw new Error('Ledger responded with errors');
+      }
 
-      // 3. Add signature to unsigned tx
+      unsignedTx.createTxWithSignature(resp.signatureVRS);
```
Comment (kyranjamie) [inline app/modals/stacking/stacking-modal.tsx:149]: This call returns a new tx without mutating the original.

```ts
const signedTx = unsignedTx.createTxWithSignature(resp.signatureVRS);
```

Comment (hstove) [reply app/modals/stacking/stacking-modal.tsx]: Thanks, fixed. If it's immutable, then I think `createTxWithSignature` makes more sense.

**Comment by kyranjamie** [comment]
Rebasing into other PR when CI finishes

**Pull request merged**